stdlib/internal/promql/join.md

```diff
@@ -0,0 +1,91 @@
+### join
+
+`join` merges two input streams into a single output stream.
+Records that have the same group key and `_time` values will be joined in the output.
+The output groups will be the same as the input groups.
+
+`join` has the following properties:
+
+| Name    | Type     | Description |
+| ------- | -------- | ----------- |
+| tables  | object   | Specifies the two table streams to join. |
+| fn      | function | Defines the function that joins the rows of each table. The return value is an object which defines the output record structure. |
+
+Ex.
+```
+import "internal/promql"
+
+left = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "A")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "a"})
+
+right = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "B")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "b"})
+
+promql.join(tables: {left, right}, fn: (t) => {
+    return {t.left with b: t.right.b}
+}
+```
+
+`join` assumes that each input table is sorted by `_time`.
+The planner will ensure this by inserting a `sort` procedure after each one of join's immediate predecessors.
+`join` will only call `fn` on rows that pass the join condition.
+
+Here are some examples of compiling PromQL queries into their equivalent Flux forms involving `promql.join`.
+
+```
+left_metric + right_metric
+```
+compiles to:
+```
+lhs = left_metric
+    |> rename(columns: {_value: "lv"})
+
+rhs = right_metric
+    |> rename(columns: {_value: "rv"})
+
+promql.join(tables: {lhs, rhs}, fn: (t) => ({t.lhs with rv: t.rv}))
```

Comment:
so the `map` is technically not needed here as we could to the `+` in the join fn.    

However, why the join fn?   we could just have `join` output  `t` for each row.  Then a call to `map` to finish it off?  

I guess it's for efficiency, yeah?

Reply:
True the addition operation could be defined within the join fn. The reason for the `fn` parameter is that it allows the user to specify the exact output schema they want which is a much more declarative design. The current join handles the merging of rows internally and any column name conflicts. I have found that to be really awkward. This way the user tells us what data they want out vs `join` giving them data that they must then reshape.

Reply:
Join shouldn't perform any processing and shouldn't act as yet another map. We cannot have them specify a schema as it won't work for the unknown schemas the transpiler works with.